modules/engine/src/main/scala/com/gsk/kg/engine/analyzer/FindUnboundVariables.scala

```diff
@@ -76,16 +78,79 @@ object FindUnboundVariables {
 
         State
           .modify[Set[VARIABLE]](x => x union vars)
-          .flatMap(_ => Set.empty.pure[ST])
+          .flatMap(_ => Set.empty[VARIABLE].pure[ST])
       case LeftJoin(l, r, filters) => (l union r).pure[ST]
       case Union(l, r)             => (l union r).pure[ST]
-      case Filter(funcs, expr)     => expr.pure[ST]
-      case Join(l, r)              => r.pure[ST]
-      case Offset(offset, r)       => r.pure[ST]
-      case Limit(limit, r)         => r.pure[ST]
-      case Distinct(r)             => r.pure[ST]
-      case Group(vars, func, r)    => r.pure[ST]
-      case DAGOrder(conds, r)      => r.pure[ST]
-      case Noop(trace)             => Set.empty.pure[ST]
+      case Filter(funcs, expr) =>
+        val funcsVars = funcs.toNes.toSortedSet
+          .foldLeft(Set.empty[VARIABLE]) { case (acc, func) =>
+            acc ++ FindVariablesOnExpression.apply[Expression](func)
+          }
+        for {
+          declared <- State.get
+        } yield (funcsVars diff declared) ++ expr
+      case Join(l, r)        => r.pure[ST]
```

Comment:
We should also fix this right?
```suggestion
      case Join(l, r)        => (l union r).pure[ST]
```

Reply:
Good catch!